### PR TITLE
Feature toggles (minus windows)

### PR DESCRIPTION
**Pull request by shakyShane**
**Reviewer:** @GioSensation 
**Asana:** https://app.asana.com/0/0/1202318756875788/f

## Description

Feature toggles + available input types on iOS + Android

Please review the separate commits as I've chunked them up to make the review easier

- [functional changes](https://github.com/duckduckgo/duckduckgo-autofill/pull/170/commits/f58eb4c5f79d883bf229022b713ba2431a23f90a)
- [integration tests](https://github.com/duckduckgo/duckduckgo-autofill/pull/170/commits/aefdd088e1b111c9ceffd5304a455552e380606a)
- [schemas](https://github.com/duckduckgo/duckduckgo-autofill/pull/170/commits/3f2d492a6aa98d1db9cb938f3ccf906dbb848dcc)

## Steps to test

Please see the asana task above for the testing scenarios covered

### CODE REVIEW

File: src/deviceApiCalls/transports/android.transport.js
Comment (GioSensation) [inline src/deviceApiCalls/transports/android.transport.js:53]: Mhm, interesting. So the messages we send from the native layer have `origin = ''`?

File: src/deviceApiCalls/transports/android.transport.js
Comment (GioSensation) [inline src/deviceApiCalls/transports/android.transport.js:43]: I think this is now incorrect. This doesn't send the message. It just waits for the response.

File: src/deviceApiCalls/transports/android.transport.js
Comment (GioSensation) [inline src/deviceApiCalls/transports/android.transport.js:33]: Just being paranoid. Is it possible that the native layer responds before we have the chance to register the listener? I don't think so, but wanted to double check with you.

File: src/deviceApiCalls/transports/apple.transport.js
Comment (GioSensation) [inline src/deviceApiCalls/transports/apple.transport.js:17]: Previously, some calls were fire-and-forget and didn't need to wait for a response (`weSend`). Did we change that?

File: src/deviceApiCalls/transports/extension.transport.js
Comment (GioSensation) [inline src/deviceApiCalls/transports/extension.transport.js:5]: This is because I suppose we haven't ported the extension calls to use the transport. Was there a specific challenge/reason for it or was it just descoped?

File: src/deviceApiCalls/transports/transports.js
Comment (GioSensation) [inline src/deviceApiCalls/transports/transports.js:31]: I think we can check for the existence of `chrome` in the global scope.

**Review by GioSensation — APPROVED** [review]
Just a bunch of comments and questions, but this is looking great. I've only smoke tested on a couple of platforms.

File: docs/runtime.ios.md
Comment (shakyShane) [reply docs/runtime.ios.md:12]: I can clarify, thank you

File: docs/runtime.android.md
Comment (shakyShane) [reply docs/runtime.android.md:165]: It's there to enable the native side to opt into 'no action' if they desire.

For example, on iOS it can be tricky to get the keyboard to show (each year there seems to be different hacks involved etc) so the native side may decide to treat a modal dismissal as either 'try to focus the field' or 'do nothing'

So, it's here to give them a choice going forward.

File: docs/settings.md
Comment (shakyShane) [reply docs/settings.md:35]: @GioSensation The idea with 'settings.refresh()' is that it may or may not include API calls, lookups from string replacements, defaults, or anything else. 

It's designed to be something that's safe to call whenever, to update 'state' of the user/application.

For example, when a user has signed into email protection it should be safe to call 'refresh()' and have it determine the next state.

Let me know if this answers your question.

File: src/DeviceInterface/AppleDeviceInterface.js
Comment (shakyShane) [reply src/DeviceInterface/AppleDeviceInterface.js:82]: @GioSensation yep - it's related to email sign-in. This can be called at any point and if signed in, we may need to re-scan and re-decorate fields

File: src/Settings.js
Comment (shakyShane) [reply src/Settings.js:60]: @GioSensation yep, an injected logger to the device is something we should do - let's make a follow up task as I didn't want to introduce any more new concepts here :)

In terms of logging in the catch, actually this is the correct flow here - the calls *will* be made on all platforms, but the transport will throw if it can't fulfil it (either by talking to the device or by reading config).

For example, see: 

https://github.com/duckduckgo/duckduckgo-autofill/blob/14c9aa25d08658ff1f83ffa4cf8eaed66ee47be3/src/deviceApiCalls/transports/android.transport.js#L91-L93

This means that the log message you mentioned above *would* give good context about the error, but only in development/test mode

File: src/deviceApiCalls/schemas/storeFormData.params.json
Comment (shakyShane) [reply src/deviceApiCalls/schemas/storeFormData.params.json:24]: @GioSensation it's only being used on Android, iOS, Windows currently, and those only support storing credentials. 

Once we begin transitioning things like the macOS calls, at that point we can expand the schema out.

File: src/deviceApiCalls/transports/transports.js
Comment (shakyShane) [reply src/deviceApiCalls/transports/transports.js:31]: That comment was left by mistake, thanks for catching

File: src/deviceApiCalls/transports/android.transport.js
Comment (shakyShane) [reply src/deviceApiCalls/transports/android.transport.js:33]: Not from what I've experienced so far on Android + Windows - once merged and Android does more testing on their end we can always circle back here. there's no risk to the existing android application because we havn't changed how the email protection calls works 🙏🏻

File: src/deviceApiCalls/transports/android.transport.js
Comment (shakyShane) [reply src/deviceApiCalls/transports/android.transport.js:53]: Yep that's correct - this is something we noticed (it's similar on windows). It appears that no page script could ever construct such an event, so this is a nice way of us filtering messages to those originating from the app.

File: src/Settings.js
Comment (GioSensation) [reply src/Settings.js:87]: Nah, I thought that this `refresh` was also legacy but apparently it isn't.

File: src/deviceApiCalls/transports/android.transport.js
Comment (GioSensation) [reply src/deviceApiCalls/transports/android.transport.js:51]: Nice!

File: src/deviceApiCalls/transports/apple.transport.js
Comment (GioSensation) [reply src/deviceApiCalls/transports/apple.transport.js:17]: Mhm, that's not entirely correct. On older devices we create methods that listen for a response (see `generateRandomMethod`). In this case, we would leave these methods there indefinitely, one for every fire-and-forget call. Is there a way to infer from the schema whether the call expects a response and if so short-circuit to `wkSend`?

Comment (shakyShane) [reply src/deviceApiCalls/transports/apple.transport.js:17]: yep, we can do that 👍🏻

**Pull request merged**
